### PR TITLE
Add TextInput validator

### LINKED ISSUE #909: Input validator
## Expected Behavior

To distribute an application one of the most important things is the validation of inputs. Due to this 
most GUI/Web frameworks provide some validator functions. 

I think validators are really missing in Toga. Currently I use the on_change attribute to call a
function which validates the input and do something based on the truth value. (https://gist.github.com/MMartin09/481bb0a5b1126c5616837bad316f7230) 
But I think there are better ways. 

For example a is_valid attribute which is the return value of a provided validate function. So I think
this could look something like this:

```python
self.text_input = toga.TextInput(validate_command=self.input_validator)

if self.text_input.is_valid:
  # do soemthing
```

Maybe there are other better ways but I think input validators are quite useful.

### PR DESCRIPTION
**Pull request by saroad2**
Initial implementation of a validator in `toga.TextInput` in Windows. One can add a `validator` parameter to the `toga.TextInput` constructor in order to validate the value. the `validator` callable should return `None` if the value is valid and an error message if it is not.

validator

closes #909

## PR Checklist:

- [X] All new features have been tested
- [X] All new features have been documented
- [X] I have read the **CONTRIBUTING.md** file
- [X] I will abide by the code of conduct

### CODE REVIEW

**Comment by saroad2** [comment]
Once the API is approved, I will tackle #378 and implement a bunch of default validators such as `toga.max_length_validator` , `toga.number_validator`, `toga.email_validator`, etc.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
The interface for setting/clearing errors makes sense, as does the mechanics of setting the validator and manually checking validity.

My concern lies with the 2 details that have been provided for Winforms, but not for any other platform:
1. The visual rendering style for validation errors
2. The mechanism for triggering validation errors

(2) isn't *that* much of a problem - worst case, I can imagine that a mechanism that validates after every input change would be easy enough to set up. A working implementation would be nice here, even if it is connected to a no-op on the actual visualization.

However (1) strikes me as a case where we're leaning pretty heavily on a Winforms API that doesn't have an immediate analog (that I can think of) in Cocoa or GTK. 

I'm entirely comfortable with committing a Windows-only solution for now, with the hope that someone will fill in the blanks later for other platforms. However, I would like to have a little more confidence in what we are intending to implement on those platforms, so that we can have a certain degree of confidence that we're not painting ourself into a corner with the API design. That doesn't need to be any more than some links to HIG design guides, or some screenshots of "authoritative" apps (e.g., an mainstream Apple-produced app demonstrating input validation).

**Comment by saroad2** [comment]
I have a great idea on how to implement this exact API in Cocoa, but it requires beeware/rubicon-objc#190. Once we figure that out, I'll show you a screenshot!

**Comment by saroad2** [comment]
OK, so I managed to do what I visioned as the implementation would be for Cocoa:

Screen Shot 2020-09-29 at 23 03 26

As you can see, when error is set, the text box is surrounded by a red border and the error is displayed in a tooltip. I think we can go with the same approach in GTK.

What do you think?

**Comment by freakboy3742** [comment]
That style of red border around bad text is definitely something I've seen on *web* forms, but I'm struggling to think of an example of a native Apple GUI that does that style of error handling. 

The closest I can see in the Apple macOS HIG is the section on [data entry](https://developer.apple.com/design/human-interface-guidelines/macos/user-interaction/data-entry/), which only says that validation should be performed on loss of focus.

As with most of these discussions, what I'm trying to capture is "What is the platform-appropriate thing to do?" There are any number of *possible* implementations; however, that won't necessarily result a GUI that feels native. I want to avoid implementing things "because we can", and make sure we're building the right thing. Examples of native apps that follow the HIG and do validation will be the most compelling argument here.

**Comment by saroad2** [comment]
Here's in example on Mac using PyCharm

Screen Shot 2020-09-30 at 14 15 51

I believe I can find more.

**Comment by freakboy3742** [comment]
So - PyCharm actually really bugs me as an app. It *doesn't* follow the HIG in so many ways. AFAICT, it *is* a web app, with a superficial macOS-looking skin (try dragging the Find dialog panel off the main window). 

The most compelling examples will be apps that Apple ships, or that Apple has given a WWDC Design Award.

**Comment by saroad2** [comment]
Here's Postman:

Screen Shot 2020-09-30 at 15 04 51

Couldn't find an example from a WWDC app.

**Comment by SamSchott** [comment]
Living on macOS, I have yet to see a visual indication of rejected input in a system app.

Whenever data entry is restricted to a subset of allowed values, system apps tend to offer a user interface which only allows the entry of correct values in the first place, instead of performing validation afterwards. Examples are:

1. Setting the host name in "System Preferences > Sharing": entering special characters like an exclamation mark is not possible.
2. Setting the date or time in a Calendar event: typing any letters will be rejected. Entering number which doesn't make sense (a date of 35/03/2020) is not possible. 
3. Any input which is restricted to a subset of selections

In some other cases, invalid input is allowed but will be reverted before leaving focus:

1. Xcode inspector in the interface builder: entering a custom class name with invalid characters.
2. Contacts: entering an invalid date of brith.

In all cases, there is an audio indication of the rejected input. The implementation likely uses [formatters](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DataFormatting/DataFormatting.html#//apple_ref/doc/uid/10000029-CJBCGJCJ):

> You use formatters to interpret and create strings that represent other data types, and to validate the text in text fields and other cells.

Its not a very consistent story and won't work well for less obvious input validations (a password which requires at least two numbers, three special characters, an upper-case letter and must have a length between 8 and 12 characters). In such cases, the macOS way would probably be to provide an explanatory text below the input and a dialog in case of invalid input.

**Comment by saroad2** [comment]
So showing an error dialog is the right approach?
I'm a little bit confused by now.

**Comment by SamSchott** [comment]
Not sure if it's the *right* approach. It's just what macOS does in case of more complicated validation requirements. Easier ones are just refused at input with an audible signal. IMO, a more uniform solution would be nice for toga.

**Comment by freakboy3742** [comment]
I agree with @SamSchott that GUIs allowing invalid input state is highly unusual in system apps - I'm definitely stretching to think of any apps that use this approach. The examples that @saroad2 has don't even appear to be using Cocoa, AFAICT. That doesn't necessarily make them *bad* GUIs - but it does mean they're not 100% *native* GUIs, which is what Toga is aiming for.

I disagree that a *uniform* solution is what we should be seeking - or at least, it should only be uniform in high-level behavior, not visual behavior or appearance. One of the goals of Toga is to represent the high level goal of a GUI, and then let the platform backend determine the "right" way to render that concept. 

The best example of this in practice at present is the menus - Mac apps have Quit at the bottom of the App menu; Windows apps have Quit at the bottom of the File menu. As a developer, you don't need to decide how/where to put Quit - the menu item appears in the right place for the platform. Another good example is the Switch widget - on some platforms, it's a checkbox; on others, it's a flick switch. The "right" widget changes between platform, depending on what the local platform style guide recommends for boolean controls - e.g, macOS says checkbox, iOS says flick switch.

In the case of this feature, the high level concept is "text input has a limited range of valid values". There are multiple ways to satisfy that requirement, and different platforms will disagree on what is platform appropriate. To my mind, it's entirely reasonable for a Windows app to render with a "this input isn't valid" border/icon, but Cocoa to reject input with a dialog (or another platform appropriate way to render errors). The end behavior is the same - content is validated against the same pattern, with the same error displayed (in whatever contextual way is appropriate), and an `is_valid()` check on the current input value can be performed. The only oddity would be that Cocoa input is *always* valid.

**Comment by saroad2** [comment]
To summarize everything up:
1. On Windows we're keeping the current implementation that I did
2. On Mac I'm going to change the behavior that the validation will occur on focus change and it will raise an error dialog instead of the red box.
3. On other platforms, we'll tackle them one-by-one in later PRs.

Sounds reasonable? Can I proceed with the implementation?

**Comment by freakboy3742** [comment]
That sounds good to me as a plan.

One implementation detail for Windows - I notice in your original rendering, the "invalid" icon falls outside the text box. How does that impact on the layout size of the box? Do we need to modify the hinted size of the box to accomodate the potential display of that icon if a validator is installed?

**Comment by saroad2** [comment]
Fixed the implementation in Cocoa to show an error dialog on validation failure and keep the value valid at all times.

In windows, moved the error icon to be ontop the text box in order to not affect the layout:

validation

File: src/dummy/toga_dummy/widgets/textinput.py [hunk 1]
```diff
@@ -28,3 +28,9 @@ def rehint(self):
 
     def set_on_change(self, handler):
         self._set_value('on_change', handler)
+
+    def set_error(self, error_message):
+        self._set_value('error', error_message)
+
+    def unset_error(self):
```
Comment (freakboy3742) [inline src/dummy/toga_dummy/widgets/textinput.py:35]: To me, "clear" would be more obvious terminology than "unset".

File: src/core/toga/widgets/textinput.py [hunk 2]
```diff
@@ -100,10 +106,40 @@ def on_change(self):
...
+
+    def is_valid(self):
+        if self.validator is None:
+            return True
+        return self.validator(self.value) is None
+
+    def set_error(self, error_message):
+        self._impl.set_error(error_message)
+
+    def unset_error(self):
```
Comment (freakboy3742) [inline src/core/toga/widgets/textinput.py:144]: I don't think `set_error` and `unset_error` should be exposed as public API. AFAICT, they're entirely internal mechanisms, and are only needed by `validate()`. 

I'd be inclined to remove these these entirely, and just invoke `self._impl.set_error()` and `self._impl.unset_error()` as required inside validate. I can't think of any reason they'd need to be overridden at the *interface* level.

File: src/cocoa/toga_cocoa/widgets/textinput.py [hunk 3]
```diff
@@ -65,3 +70,11 @@ def rehint(self):
 
     def set_on_change(self, handler):
         pass
+
+    def set_error(self, error_message):
+        self.interface.value = self.last_valid_value
+        if self.interface.window is not None:
+            self.interface.window.error_dialog("Validation Error", error_message)
+
+    def unset_error(self):
```
Comment (freakboy3742) [inline src/cocoa/toga_cocoa/widgets/textinput.py:79]: Unset error *should* be a no-op on Cocoa, as the input can never be in a truly "error" state - correct?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
In terms of the code - I've left a couple of minor comments inline, but this looks really good.

In terms of UX, I think there's a couple of issues that need to be addressed.

On Windows, it appears that the validator is only executed on focus exit. This means if you enter "foo", and get an error; then correct the error with "foo@bar", the error indicator isn't cleared until you leave focus. When the error says 
"email must contain an @", and your input *does* have an "@", I can imagine this might be confusing.

On Cocoa, I noticed two problems:

 1. If I type bad input, I wouldn't expect the input to be *cleared*. If I've typed long and complicated input and missed one character, I wouldn't expect to have to write it again.

 2. If I'm alerted to an error when an input loses focus, I'd expect the input to *regain* focus when I acknowledge the error, so that I can correct the problem. 

Effectively, I *shouldn't* be able to leave a Cocoa TextInput in an invalid state - I should be forced to either correct the input, or delete it.

**Comment by saroad2** [comment]
Fixed everything, except of the Cocoa implementation.

The way to do so is to override [this](https://developer.apple.com/documentation/appkit/nstextdelegate/1525992-textshouldendediting) method in the delegate object, but for some reason Rubicon doesn't do so. I have no idea what I'm doing wrong there.

If you could fix this real quick, I would appreciate it.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I've made the change for Cocoa's focus handling - it turns out the delegate approach we were using is now a deprecated API anyway, so it's a change we would eventually need to make. I *think* I've made the change without breaking anything, but if you're able to spot any problems, let me know.

The one last question I have is about the validation itself. The API you've implemented here is "return the error message" - which absolutely works; however, I'm wondering if "raise ValueError()" would be a more Pythonic approach? i.e., a validator function has no return value; but it raises a ValueError (wrapping the error message) if there is a validation problem. 

Thoughts?

**Comment by saroad2** [comment]
I don't think that exceptions are a good idea.

Exceptions are expensive. They make applications crash and even when they are caught, they cause performance issues because of that. Most of the times you can replace the use of exception with returning `None` and adding a None validation.

We should definitely use exceptions when we want the application to crash if something horrible is happening, something that we are certain that the app cannot run with, but we should not use them otherwise, in my opinion.

In our case, we always want to catch the validation error message and present it to the user. That means that even if we do use exceptions here, they should always be caught. Therefore I'm asking: why use exceptions in the first place?

If we do choose to use exceptions eventually, I don't think that raising `ValueError` is the right thing to do. Generally I tend to avoid raising python builtin exceptions in my code. If we choose to use exceptions, we should add our own exception class `toga.ValidationError` and raise it instead. In that way we can catch validation errors specifically and propagate any other exception.

That's my take on this. It goes without saying that I will respect any choice you would make 😊.

**Review by freakboy3742 — APPROVED** [review]
Well, I did ask :-)

I don't think exceptions are quite as expensive from a performance perspective as you suggest - they definitely are in Java, but in Python, they're quite lightweight. However, your other comments are all reasonable.

**Comment by saroad2** [comment]
Thank you for asking!

Sorry if it came out the wrong way. I try to respectful because this is your project and I don't want to force my opinion on you.

**Comment by freakboy3742** [comment]
No problem at all! I didn't read anything you said as disrespectful in the slightest, and your points were all well reasoned. I asked for an opinion, and I got one :-)

**Pull request merged**